src/System.Text.Json/src/System/Text/Json/JsonParser.cs

```diff
@@ -0,0 +1,565 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace System.Text.Json
+{
+    public struct JsonParser : IDisposable
+    {
+        private byte[] _buffer;
+        private int _index;
+        private int _end;
+        private int _dbIndex;
+        private int _insideObject;
+        private int _insideArray;
+        public JsonTokenType TokenType;
+        private bool _jsonStartIsObject;
+
+        private const int RowSize = 9;  // Do not change, unless you also change FindLocation
+
+        public enum JsonTokenType
+        {
+            // Start = 0 state reserved for internal use
+            ObjectStart = 1,
+            ObjectEnd = 2,
+            ArrayStart = 3,
+            ArrayEnd = 4,
+            Property = 5,
+            Value = 6
+        };
+
+        public enum JsonValueType
+        {
+            String,
+            Number,
+            Object,
+            Array,
+            True,
+            False,
+            Null
+        }
+
+        public JsonParser(byte[] buffer, int lengthOfJson)
+        {
+            _buffer = buffer;
+            _insideObject = 0;
+            _insideArray = 0;
+            TokenType = 0;
+            _index = 0;
+            _end = lengthOfJson;
+
+            var nextByte = _buffer[_index];
+            while (isWhiteSpace(nextByte) || nextByte == 0)
+            {
+                _index++;
+                nextByte = _buffer[_index];
+            }
+
+            _dbIndex = _end + 1;
+
+            _jsonStartIsObject = _buffer[_index] == '{';
+        }
+
+        public void Dispose()
+        {
+        }
+
+        public JsonParseObject Parse()
```

Comment:
JsonParser.Parse() relies on some of the JsonParser members. Some of those members have values set after processing the buffer passed into the JsonParser ctor such as _end and _dbIndex.